### ISSUE TITLE
Timeout does not work as expected when connecting https server via M2Crypto package

### ISSUE THREAD

**Issue by yq8925** [pywbem/pywbem#1275]
If there is timeout in pywbem.WBEMConnection, when connecting to HTTPS server via M2Crypto ssl package, the timeout is set in cim_http.py line 573: 
self.sock.set_socket_read_timeout(SSL.timeout(self.timeout))
self.sock.set_socket_write_timeout(SSL.timeout(self.timeout))

Here the timeout in WBEMConnection is second, but the value in set_socket_read_timeout is millisecond. It is a problem, we need to transfer the timeout value to ms by multiplying 1000.
self.sock.set_socket_read_timeout(SSL.timeout(self.timeout*1000))
self.sock.set_socket_write_timeout(SSL.timeout(self.timeout*1000))

As the result of this problem, if we set the timeout (e.g. 100) and the CIM server cannot answer the request quickly, the socket is timeout and cannot read data. We always got exception "TypeError: object of type 'NoneType' has no len()\n", as below
['Traceback (most recent call last):\n', '  File "C:\\Python\\lib\\site-packages\\pywbem\\cim_operations.py", line 8326, in EnumerateClasses\n    **extra)\n', '  File "C:\\Python\\lib\\site-packages\\pywbem\\cim_operations.py", line 1616, in _imethodcall\n    conn_id=self.conn_id)\n', '  File "C:\\Python\\lib\\site-packages\\pywbem\\cim_http.py", line 808, in wbem_request\n    response = client.getresponse()\n', '  File "C:\\Python\\lib\\httplib.py", line 1121, in getresponse\n    response.begin()\n', '  File "C:\\Python\\lib\\httplib.py", line 438, in begin\n    version, status, reason = self._read_status()\n', '  File "C:\\Python\\lib\\httplib.py", line 395, in _read_status\n    if len(line) > _MAXLINE:\n', "TypeError: object of type 'NoneType' has no len()\n"]

**Comment by andy-maier** [comment]
We'll look into that.
Which version of pywbem and of Python are you using?

**Comment by andy-maier** [comment]
So you say in your case the timeout expires and then does not result in a pywbem.TimeoutError, but in the TypeError you quoted?
That is quite unexpected.

I double checked that M2Crypto documents its SSL.timeout.timeout to have a first arg in seconds. If it interprets that as milliseconds, it would have a docs bug.

Can you please double check that the TypeError shows up with WBEMConnection timeout=100 without your multiplication fix, and does notsbos up with your multiplication fix?

**Comment by yq8925** [comment]
Hi Andy, 
To your questions:
1) I use pywbem 0.12.3 and python 2.7.15. 
2) SSL.timeout has a fist arg in seconds, but self.sock.set_socket_read_timeout invokes method socket.setsockopt(socket.SOL_SOCKET, socket.SO_RCVTIMEO, timeo.pack()) in which the third arg is in milliseconds. 
3) Yes, I double-checked in my env (the CIM server does not response quickly), the TypeError exception does not show with the multiplication fix.
Thanks.

**Comment by yq8925** [comment]
One more comments:
I got this problem on windows (2012R2 64bit) with both 64bit and 32bit Python.
On windows, timeout value in method winsock.setsockopt for SO_RCVTIMEO and SO_SNDTIMEO is a DWORD type in milliseconds. ( https://msdn.microsoft.com/en-us/library/windows/desktop/ms740532(v=vs.85).aspx )

On linux, the timeout value in method setsockopt is timeval structure which has the same structure with SSL.timeout. 

Looks this problem only exists on windows?

**Comment by andy-maier** [comment]
Thanks for your analysis of the problem. I agree with your judgement that the error only exists on Windows. 

On my Wundows 7 with Python 2.7.11, I have not been able to reproduce the error symptom you got. Instead, I got the following error when using pywbemcli against a mocked slow server (via [MockLab](https://app.mocklab.io/)):
```
  File "C:\Users\Andi\Envs\pywbem27\lib\site-packages\pywbem\cim_http.py", line 808, in wbem_request
    response = client.getresponse()
  File "C:\Python27\Lib\httplib.py", line 1136, in getresponse
    response.begin()
  File "C:\Python27\Lib\httplib.py", line 453, in begin
    version, status, reason = self._read_status()
  File "C:\Python27\Lib\httplib.py", line 409, in _read_status
    line = self.fp.readline(_MAXLINE + 1)
  File "C:\Python27\Lib\socket.py", line 480, in readline
    data = self._sock.recv(self._rbufsize)
  File "C:\Users\Andi\Envs\pywbem27\lib\site-packages\M2Crypto\SSL\Connection.py", line 402, in read
    return self._read_bio(size)
  File "C:\Users\Andi\Envs\pywbem27\lib\site-packages\M2Crypto\SSL\Connection.py", line 344, in _read_bio
    return m2.ssl_read(self.ssl, size, self._timeout)
M2Crypto.SSL.SSLError: (0, 'Error')
```

I have looked into the M2Crypto code and think that the fix should be done in its SSL.timeout.py module, by converting the sec,microsec arguments to a single integer value in milliseconds. I have created M2Crypto issue https://gitlab.com/m2crypto/m2crypto/issues/218 for this, and have provided M2Crypto pull request https://gitlab.com/m2crypto/m2crypto/merge_requests/196 to address this.

Could you please install M2Crypto with this fix by installing from the branch of that PR:
```
pip install git+https://gitlab.com/andy-maier/m2crypto.git@andy/fix-windows-timeout#egg=M2Crypto
```
and verify that the problem is solved for you? Your local fix of multiplying with 1000 is no longer needed then. In fact, I don't understand why that fix worked for you, because Windows setsockpt() still gets incorrect input values (two longs instead of one, and I also got the error I mentioned above when I tried your local fix).

**Comment by yq8925** [comment]
Andy, thanks for quick fix. The new SSL.timeout.py fixed my issue.

**Comment by andy-maier** [comment]
Ok, great to hear! We'll then pursue the issue with the M2Crypto team, so that the fix gets integrated into their next version. Usually, that works very well. I'll keep this issue open to track the progress on that.

**Comment by andy-maier** [comment]
Status update regarding the M2Crypto fix: The fix has been accepted and is merged into the master branch, which means the next release of the M2Crypto Python package will have it. Not clear when that is planned for.

Pywbem 0.12 or earlier will not be able to use the fix, because in these pywbem versions, the binary builds of M2Crypto are used (Python packages M2CryptoWin32/64), and they are no longer updated. Starting with the upcoming pywbem 0.13.0, the M2Crypto package is used on all OS platforms, including Windows.

**Comment by allen0817** [comment]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python2.6/site-packages/pywbem-0.10.1-py2.6.egg/pywbem/cim_operations.py", line 7023, in ExecQuery
    **extra)
  File "/usr/lib/python2.6/site-packages/pywbem-0.10.1-py2.6.egg/pywbem/cim_operations.py", line 932, in _imethodcall
    recorder=self.operation_recorder)
  File "/usr/lib/python2.6/site-packages/pywbem-0.10.1-py2.6.egg/pywbem/cim_http.py", line 756, in wbem_request
    client.endheaders()
  File "/usr/lib64/python2.6/httplib.py", line 908, in endheaders
    self._send_output()
  File "/usr/lib64/python2.6/httplib.py", line 780, in _send_output
    self.send(msg)
  File "/usr/lib/python2.6/site-packages/pywbem-0.10.1-py2.6.egg/pywbem/cim_http.py", line 446, in send
    self.connect()
  File "/usr/lib/python2.6/site-packages/pywbem-0.10.1-py2.6.egg/pywbem/cim_http.py", line 550, in connect
    ret = self.sock.connect_ssl()
  File "/usr/lib64/python2.6/site-packages/M2Crypto-0.31.0-py2.6-linux-x86_64.egg/M2Crypto/SSL/Connection.py", line 293, in connect_ssl
    return m2.ssl_connect(self.ssl, self._timeout)
M2Crypto.SSL.SSLError: (104, 'Connection reset by peer')
pywbem 0.10.1
M2Crypto 0.31
The same is true for storing versions，but I don't know if the port is 5989.Troubleshoot account password issues.
I don't know what to do

**Comment by andy-maier** [comment]
Please use pywbem 0.13.0, and not the very outdated 0.10.0.

**Comment by andy-maier** [comment]
We will release pywbem 0.13.0 today or tomorrow, and it will use the M2Crypto branch with the fix that solved your original problem. Closing this issue. If you encounter issues with pywbem 0.13.0, please open an new issue. Note that pywbem 0.10.0 is not supported anymore.

**Comment by andy-maier** [comment]
In fact, I have to correct myself. M2Crypto meanwhile released 0.31.0 to Pypi (which includes the time fix), and pywbem will pick it up as a minimum in its 0.13.0 release.
